### DISCUSSION TITLE
if () branches on unexisting process.env property aren't stripped out

### DISCUSSION THREAD

**Discussion by ohchi** [quasarframework/quasar#16004] [General - Components / Directives / etc] [answered]
Hello, dear community!

The docs say:
> When compiling your website/app, if () branches depending on process.env are evaluated, and if the expression is false, they get stripped out of the file.

And it is true as long as `process.env` property exists. But if i'm trying to make if () branch on unexisting property, it is not stripped out. I can't figure out what is the reason behind it. Is it done on purpose?

**Comment by pdanpdan** [comment]
The env constants are replaced with their value (as in a preprocessor).
The dead branch removal is done based on false.

**Reply by ohchi** [reply]
But the value inside if () branch is evaluated as falsy. It's seen from the runtime behaviour. So, let's say i have in quasar.config.js build.env:
```
FEATURE: false  // or 0 or "" or undefined
```
This causes stripping out.
But if i comment this line:
```
// FEATURE: false  // or 0 or "" or undefined
```
Stripping out stops working.

While runtime both variants work.

**Comment by pdanpdan** [comment]
If the value is not present in env it is not replaced in the first step I think

**Reply by ohchi** [reply]
Is it on purpose or, may be, "undefined" type isn't taken into account during AST tree traversal?
What do you think?

**Comment by pdanpdan** [comment]
From what you say undefined is replaced - so i suppose only the keys that are present in env are replaced.
But this is easy to deal with because you already create the env in quasar conf - și just init the key

**Reply by ohchi** [reply]
Yes, i thought about this. Not very convenient, but possible. My feature set is originated from [build root](https://buildroot.org/) profile where all disabled features look like this:
```
#
# Legacy packages
#
# BR2_PACKAGE_ACCEL_PPP is not set
# BR2_PACKAGE_DARKSTAT is not set
# BR2_PACKAGE_IGMPPROXY is not set
```
They are simply commented.

**Reply by ohchi** [reply]
No, it's impossible to init all keys for me. In build root menuconfig (menu-driven simple user interface originally used for linux kernel configuration) values which are disabled because of their dependency is disabled aren't present in the profile at all, even commented. So, the only way to deal with it is keeping track of all features in parallel with system wide profile or implementing a parser of Config.in files which are the source for menuconfig utility.

**Reply by ohchi** [reply]
But i'm starting to understand the reason why unknown keys aren't replaced with undefined values. When i build the mobile app or electron i definitely would like to have all the features inside because it should work with all the product line. In previous version of the app (not on quasar) i solved the problem by enabling preprocessor (it was babel plugin) only for spa build.

**Reply by freddy38510** [reply]
If parsing Config.in files is too unpredictable, you could write a Vite plugin to replace the missing keys by `false`. You can get inspiration from the [vite define plugin itself](https://github.com/vitejs/vite/blob/v2.9.16/packages/vite/src/node/plugins/define.ts#L69).


In fact, Quasar [passes the value of `build.env`](https://github.com/quasarframework/quasar/blob/1550db1441aa7a9eb9b432c3f5920ee2cd1cc78b/app-vite/lib/config-tools.js#L118) to the [`define`](https://v2.vitejs.dev/config/#define) option of vite.

As you can see, vite only replaces entries that come from the `define` option.

**Comment by ohchi** [comment, marked as answer]
Finally I decided not to rely upon implicit values of missing constants. Yes, it's more manual work, but i don't use everything in profile and it gives me an extra advantage:

I can benefit from typescript type checking if i initialize constants based on the types i declare (e.g. in `constants.d.ts` file).

It's worth noting that i decided to use `rawDefine` instead of `env`.
